src/network/network_admin.cpp

```diff
@@ -647,10 +678,48 @@ NetworkRecvStatus ServerNetworkAdminSocketHandler::Receive_ADMIN_JOIN(Packet *p)
 		return this->SendError(NETWORK_ERROR_ILLEGAL_PACKET);
 	}
 
-	this->status = ADMIN_STATUS_ACTIVE;
-
 	Debug(net, 3, "[admin] '{}' ({}) has connected", this->admin_name, this->admin_version);
 
+	return is_public_key ? this->SendNeedKeyAuth() : this->SendProtocol();
+}
+
+static bool AdminCheckSignatureV1(std::span<uint8_t> signature, std::string_view public_key, std::span<uint8_t> challenge)
+{
+	public_key.remove_prefix(NETWORK_PUBLIC_KEY_V1_PREAMBLE.size());
+
+	std::array<uint8_t, NETWORK_PUBLIC_KEY_V1_SIZE_BYTES> public_key_value;
+	if (public_key.size() != NETWORK_PUBLIC_KEY_V1_SIZE_BYTES * 2 ||
+			!ConvertHexToBytes(public_key, public_key_value)) {
+		return false;
+	}
+
+	auto res = crypto_ed25519_check(signature.data(), public_key_value.data(), challenge.data(), challenge.size());
+	return res == 0;
+}
+
+NetworkRecvStatus ServerNetworkAdminSocketHandler::Receive_ADMIN_KEYAUTH(Packet *p)
+{
+	if (this->status != ADMIN_STATUS_KEYAUTH) return this->SendError(NETWORK_ERROR_NOT_EXPECTED);
+
+	bool valid_response = false;
+	if (this->public_key.starts_with(NETWORK_PUBLIC_KEY_V1_PREAMBLE)) {
+		std::array<uint8_t, NETWORK_PUBLIC_KEY_SIGNATURE_LENGTH> signature;
+		for (uint8_t &s : signature) s = p->Recv_uint8();
+
+		valid_response = AdminCheckSignatureV1(signature, this->public_key, this->challenge);
+	} else {
+		/* In the ADMIN_JOIN the version has already been checked. */
+		NOT_REACHED();
+	}
+
+	const auto &allowed_public_keys = _settings_client.network.admin_public_keys;
```

Comment:
I really like you only validate the public key AFTER the client has verified it has access to the private key. That is a rather clever approach :)